### PR TITLE
Candidates filter update and create internship filter

### PR DESCRIPTION
**Pull request by ihhyy**

### CODE REVIEW

File: InternshipBE/BL/Mapping/Profiles/InternshipProfile.cs [hunk 1]
```diff
@@ -42,7 +43,8 @@ public InternshipProfile()
                 .ForMember(entity => entity.MaxCandidateCount, src => src.MapFrom(dto => dto.MaxCandidateCount))
                 .ForMember(entity => entity.RegistrationStartDate, src => src.MapFrom(dto => dto.RegistrationStartDate))
                 .ForMember(entity => entity.RegistrationFinishDate, src => src.MapFrom(dto => dto.RegistrationFinishDate))
-                .ForMember(entity => entity.LanguageType, src => src.MapFrom(dto => Enum.Parse<LanguageType>(dto.LanguageType)))
+                ///.ForMember(entity => entity.LanguageTypes, src => src.MapFrom(dto => Enum.Parse<LanguageType>(dto.LanguageTypes)))
```
Comment (romanstat) [inline InternshipBE/BL/Mapping/Profiles/InternshipProfile.cs]: comments

File: InternshipBE/BL/Services/CandidateService.cs [hunk 2]
```diff
@@ -50,18 +50,18 @@ public async Task<CandidateDTO> UpdateCandidateAsync(CandidateDTO candidate)
...
             var candidates = await _unitOfWork.Candidates.GetCandidatesByInternshipIdAsync(internshipId, pageSize, pageNumber);
 
-            if (sortBy != null)
-            {
-                candidates = SortCandidates(candidates, sortBy, desc);
-            }
+            //if (sortBy != null)
+            //{
+            //    candidates = SortCandidates(candidates, sortBy, desc);
+            //}
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: comments

File: InternshipBE/BL/Services/CandidateService.cs [hunk 3]
```diff
@@ -89,32 +89,16 @@ public async Task<List<CandidateDTO>> SearchAsync(CandidateDTO searchModel)
...
+        //    var propertyInfo = typeof(Candidate).GetProperty(sortBy);
 
-            if (desc)
-                candidates = candidates.AsEnumerable<Candidate>().OrderByDescending(c => propertyInfo.GetValue(c, null)).ToList();
-            else
-                candidates = candidates.AsEnumerable<Candidate>().OrderBy(c => propertyInfo.GetValue(c, null)).ToList();
+        //    if (desc)
+        //        candidates = candidates.AsEnumerable<Candidate>().OrderByDescending(c => propertyInfo.GetValue(c, null)).ToList();
+        //    else
+        //        candidates = candidates.AsEnumerable<Candidate>().OrderBy(c => propertyInfo.GetValue(c, null)).ToList();
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: comments

File: InternshipBE/DAL/Entities/Filtering/CandidateFilterModel.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public string? Location { get; set; }
+
+        public LanguageType? LanguageType { get; set; }
+
+        public CandidateStatusType? StatusType { get; set; }
+
+        public string? HardSkills { get; set; }
+
+        ///public int? IntershipYear { get; set; }
```
Comment (romanstat) [inline InternshipBE/DAL/Entities/Filtering/CandidateFilterModel.cs]: comment

File: InternshipBE/DAL/Entities/Filtering/IntershipFilterModel.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace DAL.Entities.Filtering
+{
+    public class IntershipFilterModel
+    {
+        public string? Location { get; set; }
+
+        public List<LanguageType>? LanguageTypes { get; set; }
+
+        ///public LanguageType? LanguageType { get; set; }
```
Comment (romanstat) [inline InternshipBE/DAL/Entities/Filtering/IntershipFilterModel.cs]: comment

File: InternshipBE/DAL/Entities/Internship.cs [hunk 6]
```diff
@@ -22,7 +23,9 @@ public class Internship
 
         public DateTime RegistrationFinishDate { get; set; }
 
-        public LanguageType LanguageType { get; set; }
+        ///public LanguageType LanguageType { get; set; }
```
Comment (romanstat) [inline InternshipBE/DAL/Entities/Internship.cs]: comment

File: InternshipBE/DAL/Interfaces/IInternshipRepository.cs [hunk 7]
```diff
@@ -7,5 +8,7 @@ namespace DAL.Interfaces
     public interface IInternshipRepository : IGenericRepository<Internship>
     {
         Task<List<Internship>> GetInternshipsAsync(int pageSize, int pageNumber);
+
+        Task<List<Internship>> GetFilteredInternships(IntershipFilterModel filterBy, int pageSize, int pageNumber);
```
Comment (romanstat) [inline InternshipBE/DAL/Interfaces/IInternshipRepository.cs]: Async

File: InternshipBE/DAL/Repositories/CandidateRepository.cs [hunk 8]
```diff
@@ -69,11 +70,24 @@ public async Task<List<Candidate>> SearchCandidatesAsync(int skip, int take, str
             else return null;
         }
 
-        public IQueryable<Candidate> GetCandidatesForFIlter()
+        public async Task<List<Candidate>> GetCandidatesForFIlter(CandidateFilterModel filterBy)
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/CandidateRepository.cs]: FIlter -> Filter

File: InternshipBE/WebApi/Models/AGGridFilterRequest.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace WebApi.Models
+{
+    public class AGGridFilterRequest
+    {
+        public int PageNumber { get; set; }
+
+        public int PageSize { get; set; }
+
+        public IntershipFilterModel IntershipsFilterBy { get; set; }
```
Comment (romanstat) [inline InternshipBE/WebApi/Models/AGGridFilterRequest.cs]: AGGridFilterRequest contains IntershipFilterModel?

File: InternshipBE/DAL/Repositories/InternshipRepository.cs [hunk 10]
```diff
@@ -37,5 +38,34 @@ public async Task<List<Internship>> GetInternshipsAsync(int pageSize, int pageNu
                                              .Take(pageSize)
                                              .ToListAsync();
         }
+
+        public async Task<List<Internship>> GetFilteredInternships(IntershipFilterModel filterBy, int pageSize, int pageNumber)
+        {
+            var interships = await _context.Internships.ToListAsync();
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/InternshipRepository.cs]: IEnumerable?

File: InternshipBE/DAL/Repositories/InternshipRepository.cs [hunk 11]
```diff
@@ -37,5 +38,34 @@ public async Task<List<Internship>> GetInternshipsAsync(int pageSize, int pageNu
...
+                {
+                    interships = interships.Where(i => i.InternshipStacks.Any(s => s.TechnologyStackType == stack)).ToList();
+                }
+            if (filterBy.AttachedUsers != null)
+                foreach (var userName in filterBy.AttachedUsers)
+                {
+                    interships = interships.Where(i => i.Users.Any(u => u.UserName == userName)).ToList();
+                }
+            if(filterBy.IntershipYear.HasValue)
+                interships = interships.Where(i => i.StartDate.Year == filterBy.IntershipYear).ToList();
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/InternshipRepository.cs]: iqueryable

File: InternshipBE/DAL/Repositories/CandidateRepository.cs [hunk 12]
```diff
@@ -69,11 +70,24 @@ public async Task<List<Candidate>> SearchCandidatesAsync(int skip, int take, str
...
+            if (filterBy.LanguageType.HasValue)
+                candidates = candidates.Where(c => c.InternshipLanguage == filterBy.LanguageType).ToList();
+            if (filterBy.StatusType.HasValue)
+                candidates = candidates.Where(c => c.StatusType == filterBy.StatusType).ToList();
+            if(filterBy.HardSkills != null)
+                candidates = candidates.Where(c => c.PrimarySkill == filterBy.HardSkills).ToList();
+            if(filterBy.EnglishLevel.HasValue)
+                candidates = candidates.Where(c => c.EnglishLevelType == filterBy.EnglishLevel).ToList();
+            if (filterBy.UserId != null)
+                candidates = candidates.Where(c => c.Users.Any(u => u.Id == filterBy.UserId)).ToList();
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/CandidateRepository.cs]: iqueryable

File: InternshipBE/BL/DTOs/CandidateDTO.cs [hunk 13]
```diff
@@ -5,7 +5,7 @@
```
Comment (romanstat) [inline InternshipBE/BL/DTOs/CandidateDTO.cs:4]: merge conflicts

File: InternshipBE/BL/Mapping/Profiles/InternshipProfile.cs [hunk 14]
```diff
@@ -26,7 +26,8 @@ public InternshipProfile()
                 .ForMember(dto => dto.MaxCandidateCount, src => src.MapFrom(entity => entity.MaxCandidateCount))
                 .ForMember(dto => dto.RegistrationStartDate, src => src.MapFrom(entity => entity.RegistrationStartDate))
                 .ForMember(dto => dto.RegistrationFinishDate, src => src.MapFrom(entity => entity.RegistrationFinishDate))
-                .ForMember(dto => dto.LanguageType, src => src.MapFrom(entity => entity.LanguageType.ToString()))
+                ///.ForMember(dto => dto.LanguageTypes, src => src.MapFrom(entity => entity.LanguageTypes))
```
Comment (romanstat) [inline InternshipBE/BL/Mapping/Profiles/InternshipProfile.cs]: comment

File: InternshipBE/BL/Services/InternshipService.cs [hunk 15]
```diff
@@ -29,10 +30,13 @@ public async Task<InternshipDTO> CreateInternshipAsync(InternshipDTO newInternsh
             return _mapper.Map<InternshipDTO>(internship);
         }
 
-        public async Task<List<InternshipDTO>> GetInternshipsAsync(int pageSize, int pageNumber)
+        public async Task<List<InternshipDTO>> GetInternshipsAsync(int pageSize, int pageNumber, IntershipFilterModel filterBy)
         {
             var internships = await _unitOfWork.Internships.GetInternshipsAsync(pageSize, pageNumber);
 
+            if (filterBy != null)
+                internships = await _unitOfWork.Internships.GetFilteredInternshipsAsync(filterBy, pageSize, pageNumber);
```
Comment (romanstat) [inline InternshipBE/BL/Services/InternshipService.cs]: get internships than get again and replace, optimize

File: InternshipBE/DAL/Repositories/InternshipRepository.cs [hunk 16]
```diff
@@ -37,5 +38,40 @@ public async Task<List<Internship>> GetInternshipsAsync(int pageSize, int pageNu
...
+                {
+                    internships = internships.Where(i => i.InternshipStacks.Any(s => s.TechnologyStackType == stack));
+                }
+            if (filterBy.AttachedUsers != null)
+                foreach (var userName in filterBy.AttachedUsers)
+                {
+                    internships = internships.Where(i => i.Users.Any(u => u.UserName == userName));
+                }
+            if(filterBy.IntershipYear.HasValue)
+                internships = internships.Where(i => i.StartDate.Year == filterBy.IntershipYear);
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/InternshipRepository.cs]: ?

**Review by romanstat — APPROVED** [review]

**Review by vldmrklnk — APPROVED** [review]

**Review by Nikityanov — APPROVED** [review]

**Pull request merged**